### PR TITLE
Vue dependency update, better selection performance, visible quotes in text inputs

### LINKED ISSUE #9161: Render quote characters outside of the text pills
From:

image

To:

image

But make them the correct font, colour, position (not my horribly drawn red quotes :) )

### PR DESCRIPTION
**Pull request by Frizi**
### Pull Request Description

- Improved performance by batching simulatenous node edits, including metadata updates when dragging many selected nodes together.
- Updated Vue to new version, allowing us to use `defineModel`.
- Fixed #9161
- Unified all handling of auto-blur by making `useAutoBlur` cheap to register - all logic goes through a single window event handler.
- Combined all `ResizeObserver`s into one.
- Fixed the behaviour of repeated toast messages. Now only the latest compilation status is visible at any given time, and the errors disappear once compilation passes.
- Actually fixed broken interaction of node and visualization widths. There no longer is a style feedback loop and the visible node backdrop width no longer jumps or randomly fails to update.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed, the GUI was tested when built using `./run ide build`.

### CODE REVIEW

File: app/gui2/shared/ast/tree.ts [hunk 1]
```diff
@@ -1222,10 +1222,35 @@ export class TextLiteral extends Ast {
...
+        ?.join() ?? ''
+    )
+  }
+
+  get open(): Token | undefined {
+    return this.module.getToken(this.fields.get('open')?.node)
+  }
+
+  get close(): Token | undefined {
+    return this.module.getToken(this.fields.get('open')?.node)
```
Comment (kazcw) [inline app/gui2/shared/ast/tree.ts]:
```suggestion
    return this.module.getToken(this.fields.get('close')?.node)
```

File: app/gui2/src/composables/navigator.ts [hunk 2]
```diff
@@ -4,7 +4,7 @@ import { useApproach } from '@/composables/animation'
 import { PointerButtonMask, useEvent, usePointer, useResizeObserver } from '@/composables/events'
 import { Rect } from '@/util/data/rect'
 import { Vec2 } from '@/util/data/vec2'
-import { computed, proxyRefs, ref, type Ref } from 'vue'
+import { computed, proxyRefs, ref as shallowRef, type Ref } from 'vue'
```
Comment (kazcw) [inline app/gui2/src/composables/navigator.ts]: Is `ref as shallowRef` intentional? It seems misleading

File: app/gui2/src/composables/stackNavigator.ts [hunk 3]
```diff
@@ -60,7 +60,7 @@ export function useStackNavigator() {
         }
       }
     }
-    graphStore.updateState()
+    // graphStore.updateState()
```
Comment (kazcw) [inline app/gui2/src/composables/stackNavigator.ts]: Commented code? (and several similar instances below)

**Comment by farmaazon** [comment]
image

:open_mouth: 

But most of it is package.lock I guess.

File: app/gui2/src/components/GraphEditor/GraphVisualization.vue [hunk 4]
```diff
@@ -46,6 +46,7 @@ const props = defineProps<{
 }>()
 const emit = defineEmits<{
   'update:rect': [rect: Rect | undefined]
+  'update:userSetSize': [size: Vec2]
```
Comment (farmaazon) [inline app/gui2/src/components/GraphEditor/GraphVisualization.vue]: This could be a v-model, couldn't it? Now we keep two refs representing userSetSize (here and in GraphNode).

File: app/gui2/src/components/GraphEditor/NodeWidget.vue [hunk 5]
```diff
@@ -43,16 +40,17 @@ const sameInputParentWidgets = computed(() =>
...
     {
       input: props.input,
       nesting: nesting.value,
     },
     sameInputParentWidgets.value,
   )
-})
-
+}, getCurrentInstance())
+watchEffect(() => updateSelection())
```
Comment (farmaazon) [inline app/gui2/src/components/GraphEditor/NodeWidget.vue:53]: Why is it better? I cannot find docs of withCtx in google.

File: app/gui2/src/components/GraphEditor/widgets/WidgetText.vue [hunk 6]
```diff
@@ -1,23 +1,46 @@
...
-import { asNot } from '@/util/data/types'
+import { MutableModule } from '@/util/ast/abstract'
 import { computed } from 'vue'
 
 const props = defineProps(widgetProps(widgetDefinition))
-const value = computed({
+const graph = useGraphStore()
+
+const inputTextLiteral = computed((): Ast.TextLiteral | undefined => {
+  if (props.input instanceof Ast.TextLiteral) return props.input
```
Comment (farmaazon) [inline app/gui2/src/components/GraphEditor/widgets/WidgetText.vue]: WidgetInput type cannot be Ast.TextLiteral. Do you mean props.input.value?

File: app/gui2/src/components/widgets/AutoSizedInput.vue
Comment (farmaazon) [inline app/gui2/src/components/widgets/AutoSizedInput.vue:1]: We don't auto-escape the text literal anymore. Now I don't know what happens if user put's `'` inside text - probably AST change will crash, failing to fix itself by adding `()`

File: app/gui2/ydoc-server/index.ts [hunk 7]
```diff
@@ -21,7 +21,7 @@ type ConnectionData = {
   user: string
 }
 
-export async function createGatewayServer(httpServer: Server, rustFFIPath: string) {
+export async function createGatewayServer(httpServer: Server, rustFFIPath?: string) {
```
Comment (farmaazon) [inline app/gui2/ydoc-server/index.ts]: Why this change? Anyone running server will do that probably in node, and in that case it won't work without path to rustFFI wasm.

**Review by farmaazon — CHANGES_REQUESTED** [review]
We silently lost escaping strings literals in WidgetText.

File: app/gui2/src/components/GraphEditor/GraphVisualization.vue (hunk 4 above)
Comment (Frizi) [reply app/gui2/src/components/GraphEditor/GraphVisualization.vue]: You will have two refs either way. This isn't really bidirectional, since you indentionally can't set that size from the GraphNode component, it is effectively read-only. Thus only an emit, but not prop. v-model would be both.

File: app/gui2/src/components/GraphEditor/NodeWidget.vue (hunk 5 above)
Comment (Frizi) [reply app/gui2/src/components/GraphEditor/NodeWidget.vue:53]: It was a necessary change after vue update. Otherwise any context usage (i.e. injects) within `registry.select` fails, because it is executed outside of component context. This change here just made it behave as it it used to in this specific instance. I couldn't really find a good way to avoid using those injects in the first place.

File: app/gui2/ydoc-server/index.ts (hunk 7 above)
Comment (Frizi) [reply app/gui2/ydoc-server/index.ts]: This is just so the type error goes away. The usage in `gatewayServer` function in vite config is not using that parameters. Since it is truly optional, I just made the types match reality.

File: app/gui2/src/components/GraphEditor/widgets/WidgetText.vue (hunk 6 above)
Comment (Frizi) [reply app/gui2/src/components/GraphEditor/widgets/WidgetText.vue]: Yeah. It seems it worked and passed my testing only because the "Ast.parse" fallback below.

File: app/gui2/src/composables/navigator.ts (hunk 2 above)
Comment (Frizi) [reply app/gui2/src/composables/navigator.ts]: Absolutely not, good eye. This must have been some kind of auto-import or refactor tool mishap.

File: app/gui2/src/components/GraphEditor/widgets/WidgetText.vue [hunk 8]
```diff
@@ -36,13 +59,29 @@ export const widgetDefinition = defineWidget(WidgetInput.isAstOrPlaceholder, {
 </script>
 
 <template>
-  <!-- See comment in GraphNode next to dragPointer definition about stopping pointerdown -->
-  <EnsoTextInputWidget v-model="value" class="WidgetText r-24" @pointerdown.stop />
+  <label class="WidgetText r-24" @pointerdown.stop>
+    <NodeWidget :input="WidgetInput.FromAst(inputTextLiteral?.open ?? emptyTextLiteral.open!)" />
+    <AutoSizedInput v-model.lazy="textContents" />
+    <NodeWidget :input="WidgetInput.FromAst(inputTextLiteral?.close ?? emptyTextLiteral.close!)" />
```
Comment (kazcw) [inline app/gui2/src/components/GraphEditor/widgets/WidgetText.vue]: I think this could display mismatched open/close tokens in some cases, like:

`a = "unclosed`

The text literal implicitly closes at EOL--I think this PR would display the text as `"unclosed'` until it is edited.

```
a = '''
    multiline
```

The opener is `'''`, and there is no closer; I think this would display it as `'''multiline'` (there are probably other issues with multiline text though--I don't think that's been addressed specifically yet).

Comment (Frizi) [reply app/gui2/src/components/GraphEditor/widgets/WidgetText.vue]: Multiline is indeed not really handled yet. But this PR doesn't really change that. I see the issue with mismatched open and close tag. I should only be using the default "empty literal" case if there is no input literal to begin with.

Comment (kazcw) [reply app/gui2/src/components/GraphEditor/widgets/WidgetText.vue]: > I should only be using the default "empty literal" case if there is no input literal to begin with.

I think it can be reached if there is an input literal, but it doesn't have a close token, as in `a = "foo`. It's a weird edge case, but it's easy to handle nicely: for single-line text we could simply display the open token as the close token, whether or not the original source has a close token.

Thinking about it more though, I don't see why we display the interpolated/uninterpolated-quotes distinction to the user. As soon as any edit occurs we are converting it to interpolated. Until/unless we will support editing uninterpolated text as uninterpolated text, shouldn't we display all text literals as if they were coded as interpolated literals?

File: app/gui2/src/components/widgets/AutoSizedInput.vue
Comment (Frizi) [reply app/gui2/src/components/widgets/AutoSizedInput.vue:1]: Fixed now. Performing the escape/unescape is no longer the responsibility of the input itself, but the AST that the text is assigned to or read from.

**Review by somebody1234 — APPROVED** [review]
approving the 1-line change in the dashboard `package.json`

File: app/gui2/src/components/GraphEditor/widgets/WidgetText.vue (hunk 8 above)
Comment (Frizi) [reply app/gui2/src/components/GraphEditor/widgets/WidgetText.vue]: My intention was to preserve the the quotes that are used in the code for as long as possible. When you edit an uninterpolated text, as long as the new contents don't contain illegal character (the quote itself), it will stay as uninterpolated.

I've now implemented your suggestion, and in case of omitted close token, the open token is displayed twice.

I'm still going to mostly ignore handling of multi-line blocks. Once we know how we want to deal with both blocks and multi-line strings in nodes in general, we can revisit what needs to change here. I believe the current implementation will strip all newline characters on edit, since this is the behavior of html input fields.

File: app/gui2/src/components/SelectionBrush.vue [hunk 9]
```diff
@@ -40,7 +40,11 @@ const brushStyle = computed(() => {
 </script>
 
 <template>
-  <div class="SelectionBrush" :class="{ hidden }" :style="brushStyle"></div>
+  <div
+    class="SelectionBrush"
+    :class="{ cursor: props.anchor == null, hidden }"
```
Comment (farmaazon) [inline app/gui2/src/components/SelectionBrush.vue]: Why do we need class for cursor? It looks like bringing back our own cursor instead of using system one.

File: app/gui2/shared/ast/tree.ts [hunk 10]
```diff
@@ -1164,22 +1165,30 @@ export interface MutableImport extends Import, MutableAst {
...
-function escape(string: string) {
-  return string.replace(/[\0\b\f\n\r\t\v"'`]/g, (match) => mapping[match]!)
+ * NOT USABLE to insert into raw strings. Does not include double-quotes (`"`). */
+export function escapeInterpolation(string: string) {
+  return string.replace(/[\0\b\f\n\r\t\v'`]/g, (match) => escapeMapping[match]!)
+}
+
+/** Interpret all escaped characters from an interpolated (`''`) Enso string. */
+export function applyInterpolation(string: string) {
+  return string.replace(/\\[0bfnrtv']|``/g, (match) => unescapeMapping[match]!)
```
Comment (farmaazon) [inline app/gui2/shared/ast/tree.ts]: I wouldn't know what this function is doing from this name. I'd rather expect some replacing parts in backticks.

File: app/gui2/shared/ast/interpolation.ts
Comment (farmaazon) [inline app/gui2/shared/ast/interpolation.ts:1]: I liked old `text` name more, as something containing utilities for text literals in AST.

Also, I repeat something I think I wrote before: I liked `unescape` much more than `applyInterpolation`. Anyone not accustomed to our code, looking for code removing escaped sequences, will search for something containing `escape` words, not `interpolation`.

File: app/gui2/shared/ast/tree.ts [hunk 11]
```diff
@@ -1398,17 +1378,27 @@ export class MutableTextLiteral extends TextLiteral implements MutableAst {
...
+  setContentUninterpolated(rawText: string) {
+    let boundary = this.boundaryTokenCode()
+    const isInterpolated = this.isInterpolated()
+    const mustBecomeInterpolated = !isInterpolated && (!boundary || rawText.includes(boundary))
+    if (mustBecomeInterpolated) {
+      boundary = "'"
+      this.setBoundaries(boundary)
+    }
+    const literalContents =
+      isInterpolated || mustBecomeInterpolated ? escapeInterpolation(rawText) : rawText
```
Comment (farmaazon) [inline app/gui2/shared/ast/tree.ts]: If the text is uninterpolated (so, it is already escaped), why do we escape it again?

Comment (farmaazon) [reply app/gui2/shared/ast/tree.ts]: Oh, it means "not meant to be interpolated" not "not yet interpolated".

File: app/gui2/shared/ast/tree.ts [hunk 12]
```diff
@@ -1338,19 +1324,19 @@ export class TextLiteral extends Ast {
...
-      const safeText = rawText.replaceAll(/[^-+A-Za-z0-9_. ]/, '')
+      const safeText = rawText.replaceAll(/[^-+A-Za-z0-9_. ]/g, '')
       return this.new(safeText, module)
     }
     return parsed
   }
 
   /** Return the value of the string, interpreted except for any interpolated expressions. */
-  contentUninterpolated(): string {
+  get contentUninterpolated(): string {
```
Comment (farmaazon) [inline app/gui2/shared/ast/tree.ts]: But then, I'm confused here. What does it mean to "get text not meant to be interpolated from text meant to be interpolated"?

**Review by farmaazon — APPROVED** [review]
Codewise I think it's ok, but the naming I find very confusing.

I would stick with the following terms:
* interpolated/raw: one is a string _meant_ to be interpolated, second is not.
* escaped/unescaped: introduce or substitute escapes in strings, both in interpolated and raw strings. The unescaping includes any `\uXX` as those are "escape sequences".
* applying interpolation is substituting both escapes and enso code snippets with their value - something we don't do.

We may discuss it on some call.

File: app/gui2/e2e/widgets.spec.ts [hunk 13]
```diff
@@ -160,6 +161,8 @@ test('Managing aggregates in `aggregate` node', async ({ page }) => {
     'Aggregate_Column',
     '.',
     'Count_Distinct',
+    "'",
+    "'",
```
Comment (farmaazon) [inline app/gui2/e2e/widgets.spec.ts]: I would somehow make this check checking if a new parameter for Count_Distinct appeared.

File: app/gui2/playwright.config.ts [hunk 14]
```diff
@@ -45,7 +45,7 @@ export default defineConfig({
   },
   use: {
     headless: !DEBUG,
-    trace: 'on-first-retry',
+    trace: 'on',
```
Comment (farmaazon) [inline app/gui2/playwright.config.ts]: On develop it should be "retain-on-failure"

File: app/gui2/e2e/widgets.spec.ts (hunk 13 above)
Comment (Frizi) [reply app/gui2/e2e/widgets.spec.ts]: It is. The input contents is checked below. The change is that the quotes are actually rendered as tokens, not as the input contents.

File: app/gui2/playwright.config.ts (hunk 14 above)
Comment (Frizi) [reply app/gui2/playwright.config.ts]: That got resolved during the merge. It is as you say now.
